pkg/agent/controller/egress/ip_scheduler.go

```diff
@@ -0,0 +1,281 @@
+	crdv1a2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlisters "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+)
+
+const (
+	// workItem is the only item that will be enqueued, used to trigger Egress IP scheduling.
+	workItem = "key"
+)
+
+// scheduleEventHandler is a callback when an Egress is rescheduled.
+type scheduleEventHandler func(egress string)
+
+// scheduleResult is the schedule result of an Egress, including the effective Egress IP and Node.
+type scheduleResult struct {
+	ip   string
+	node string
+}
+
+// egressIPScheduler is responsible for scheduling Egress IPs to appropriate Nodes according to the Node selector of the
+// IP pool, taking Node's capacity into consideration.
+// The scheduling process takes the spec of Egress and ExternalIPPool and the state of memberlist cluster as inputs,
+// generates scheduling results deterministically. It will be triggerred when any schedulable Egress changes or the
+// memberlist cluster changes.
+type egressIPScheduler struct {
+	// cluster is responsible for selecting a Node for a given IP and pool.
+	cluster memberlist.Interface
+
+	egressLister       crdlisters.EgressLister
+	egressListerSynced cache.InformerSynced
+
+	// queue is used to trigger scheduling. Triggering multiple times before the item is consumed will only cause one
+	// execution of scheduling.
+	queue workqueue.Interface
+
+	// mutex is used to protect scheduleResults.
+	mutex           sync.RWMutex
+	scheduleResults map[string]*scheduleResult
+	// scheduledOnce indicates whether scheduling has been executed at lease once.
+	scheduledOnce *atomic.Bool
+
+	// eventHandlers is the registered callbacks.
+	eventHandlers []scheduleEventHandler
+
+	// The maximum number of Egress IPs a Node can accommodate.
+	maxEgressIPsPerNode int
+}
+
+func NewEgressIPScheduler(cluster memberlist.Interface, egressInformer crdinformers.EgressInformer, maxEgressIPsPerNode int) *egressIPScheduler {
+	s := &egressIPScheduler{
+		cluster:             cluster,
+		egressLister:        egressInformer.Lister(),
+		egressListerSynced:  egressInformer.Informer().HasSynced,
+		scheduleResults:     map[string]*scheduleResult{},
+		scheduledOnce:       &atomic.Bool{},
+		maxEgressIPsPerNode: maxEgressIPsPerNode,
+		queue:               workqueue.New(),
+	}
+	egressInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.addEgress,
+			UpdateFunc: s.updateEgress,
+			DeleteFunc: s.deleteEgress,
+		},
+		resyncPeriod,
+	)
+
+	s.cluster.AddClusterEventHandler(func(poolName string) {
+		// Trigger scheduling regardless of which pool is changed.
+		s.queue.Add(workItem)
+	})
+	return s
+}
+
+// addEgress processes Egress ADD events.
+func (s *egressIPScheduler) addEgress(obj interface{}) {
+	egress := obj.(*crdv1a2.Egress)
+	if !isEgressSchedulable(egress) {
+		return
+	}
+	s.queue.Add(workItem)
```

Comment:
Curious - Is it possible to only process the new Egress ADD events without re-calculating the previous results?  Maybe it is not worth doing so as it may bring extra complexity.

Reply:
It's not very worth as Egress ADD events don't come with an Egress IP normally unless user specifies one when creating it. It's normally the UPDATE event carrying an IP, but it's not easy to know whether this Egress would affect others without looking at the entire list.

Reply:
Got it. Thanks for the explanation.